Description: cogl: Restore support for EGL 1.4 and GLES 2.0
 This is required for Xilinx (Mali) hardware support.
Author: Talha Can Havadar <talha.can.havadar@canonical.com>
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4021
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/mutter/+bug/2080602
Last-Update: 2024-09-30
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/cogl/cogl/driver/gl/gl/cogl-driver-gl.c
+++ b/cogl/cogl/driver/gl/gl/cogl-driver-gl.c
@@ -431,6 +431,13 @@
   ctx->glGetString =
     (void *) _cogl_renderer_get_proc_address (ctx->display->renderer,
                                               "glGetString");
+
+  if (!check_gl_version (ctx, gl_extensions, error))
+    return FALSE;
+
+  /* These are only used in _cogl_context_get_gl_extensions for GL 3.0
+   * so don't look them up before check_gl_version()
+   */
   ctx->glGetStringi =
     (void *) _cogl_renderer_get_proc_address (ctx->display->renderer,
                                               "glGetStringi");
@@ -440,9 +447,6 @@
 
   gl_extensions = _cogl_context_get_gl_extensions (ctx);
 
-  if (!check_gl_version (ctx, gl_extensions, error))
-    return FALSE;
-
   if (G_UNLIKELY (COGL_DEBUG_ENABLED (COGL_DEBUG_WINSYS)))
     {
       char *all_extensions = g_strjoinv (" ", gl_extensions);
--- a/cogl/cogl/driver/gl/gles/cogl-driver-gles.c
+++ b/cogl/cogl/driver/gl/gles/cogl-driver-gles.c
@@ -580,9 +580,6 @@
   context->glGetString =
     (void *) _cogl_renderer_get_proc_address (context->display->renderer,
                                               "glGetString");
-  context->glGetStringi =
-    (void *) _cogl_renderer_get_proc_address (context->display->renderer,
-                                              "glGetStringi");
 
   gl_extensions = _cogl_context_get_gl_extensions (context);
 
--- a/cogl/cogl/winsys/cogl-winsys-egl.c
+++ b/cogl/cogl/winsys/cogl-winsys-egl.c
@@ -113,7 +113,12 @@
 _cogl_winsys_renderer_get_proc_address (CoglRenderer *renderer,
                                         const char   *name)
 {
-  return eglGetProcAddress (name);
+  GCallback result = eglGetProcAddress (name);
+
+  if (result == NULL)
+    g_module_symbol (renderer->libgl_module, name, (gpointer *)&result);
+
+  return result;
 }
 
 static void
